src/erl_optics_server.erl

```diff
@@ -0,0 +1,118 @@
+-module(erl_optics_server).
+-include("erl_optics.hrl").
+
+-behaviour(gen_server).
+
+-export([init/1,
+         handle_call/3,
+         handle_cast/2,
+         handle_info/2,
+         terminate/2,
+         code_change/3]).
+
+-export([start_link/1,
+         stop/0]).
+
+%for prototyping only
+-export([start/0,
+         test_update/0,
+         start_test/1]).
+
+-define(SERVER, ?MODULE).
+
+-record(state, {port = ?ENV_PORT     :: inet:port_number(),
+                addr = ?ENV_HOSTNAME :: inet:socket_address() | inet:hostname(),
+                mode                 :: carbon | prometheus | erlang}).
+
+
+%%%=========
+%%% API
+%%%=========
+
+%Modes: prometheus | carbon | erlang
+
+start_link(Mode) ->
+    gen_server:start_link({local, ?SERVER}, ?MODULE, [Mode], []).
+
+stop() ->
+    gen_server:cast(?SERVER, stop).
+
+start_test(Interval) ->
+    gen_server:cast(?SERVER, {test_update, Interval}).
+
+%%%==========
+%%% Callbacks
+%%%==========
+
+init([Mode]) ->
+    case Mode of
+        carbon ->
+            Hostname = ?ENV(?ENV_HOSTNAME, ?DEFAULT_HOSTNAME),
+            Port = ?ENV(?ENV_PORT, ?DEFAULT_PORT),
+            Interval = ?ENV(?ENV_INTERVAL, ?DEFAULT_INTERVAL),
+            erl_optics:allocate_carbon_poller(Hostname, Port),
+            timer:send_interval(Interval, carbon_poll),
+            %gen_server:cast(?SERVER, {test_update, 10}), %for testing
+            {ok, #state{mode = carbon}, 0};
+        prometheus ->
+            %to be implemented
+            {ok, #state{mode = prometheus}, 0};
+        erlang ->
+            %not very useful by itself
+            Interval = ?ENV(?ENV_INTERVAL, ?DEFAULT_INTERVAL),
+            timer:send_interval(Interval, erlang_poll),
+            {ok, #state{mode = erlang}, 0}
+    end.
+
+handle_call(erlang_poll, _From, State) ->
+    {reply, erl_optics:poll(), State}.
+
+handle_cast(stop, State) ->
+    {stop, normal, State};
+handle_cast({test_update, Interval}, State) ->
+    timer:send_interval(Interval, test_update),
+    {noreply, State}.
+
+handle_info(timeout, State) ->
+    {noreply, State};
+handle_info(test_update, State) ->
+    test_update(),
+    {noreply, State};
+handle_info(carbon_poll, State) ->
```

Comment:
Having no function head that matches any message leaves you open to `function_clause` exceptions.